RFC for CouchDB background workers

nickva opened the pull request.

garrensmith commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,252 @@
+If a user decided to cancel a job that is running, they MUST toggle `CancelReq`
+to `true`. Then they MUST wait for the job to stop running. If the job is not
+running, they may directly remove the job from the pending and the main jobs
+table.
```
What happens in the case of very long running jobs? e.g continuous replication or indexing a very large db. If we get into a situation where the replication queue is has more continuous replications and than workers, some jobs will never be processed. I think we would have to look at having the workers periodically check the queue and decide if they need to pause their job, add it back to the pending queue with new options and pick another job.

nickva replied:
Exactly. Yeah, in case of replications something like that might be needed and we might preserve some of the previous scheduler features, like you pointed out:

```
  If there is a pending backlog:
     Stop `$max_churn` number of continuous jobs put them back on the queue
     Pick up `$max_churn` number of jobs from the queue and run those 
```

Good idea, I'll add this to the RFC.

nickva commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,252 @@
+ - `("couch_workers", "jobs", JobType, JobId) = (JobState, WorkerId, Priority, CancelReq, JobInfo, JobOps)`
+
+Pending queue:
+ - `("couch_workers", "pending", JobType, Priority, JobId) = ""`
```
I was thinking the priority value would encode the order of the queue. It could be just an integer, a coarse timestamp (sometimes in the next 5 minutes), a tag with maybe a string.

The JobId would be the random value at the end to avoid contention as long as JobIds are unique. If they are not then a random value should be added to the Priority perhaps.

nickva commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,252 @@
+Pending queue:
+ - `("couch_workers", "pending", JobType, Priority, JobId) = ""`
+
+Active queue:
```
You're right. Thanks, it's just table in this case.

nickva commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,252 @@
+using it to implement the `/_active_tasks` API endpoint.
+
+When the HTTP API is access, a single `get_range(...)` request for
+`("couch_workers", "active")` subspace MAY be used to generate the results.
```
Yeah that is a good idea. I couldn't think of a good shape for the API yet and left it out. Once we get a feel for it, (play with a prototype), it might be easier to see what it might look like.

nickva commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,252 @@
+Worker registration and health:
+ - `("couch_workers", WorkerType, "workers_vs") = Versionstamp`
+ - `("couch_workers", WorkerType, "workers", WorkerId) = WorkerOpts`
+ - `("couch_workers", WorkerType, "health", WorkerId) = (Versionstamp, Timestamp, WorkerTimeout)`
```
Good point. I had added `Versionstamp` in there for that reason to allow monitors to look at the times between `versionstamp` changes. The timestamp might be useful for diagnostics in the sense that we'd like to know when was the last time it was updated.  But I think in that case maybe we'd just let jobs maintain their own last update timestamp history and this should just be an atomic counter like you say.

garrensmith commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,287 @@
+
+### Liveliness Race Condition
+
+In case workers come back alive after not updating their health entry often
```
Maybe we should reword this to when a worker starts up and during each health update, they must check that their jobs haven't been returned to the queue.

nickva replied:
I added that workers should check if they are still part of the active workers membership list (they can listen to the workers_vs viewstamp with a watch). If all of the sudden they find themselves removed from the list they should stop running all their jobs, re-initialize and go back to the job queue to pick new jobs.

sansato commented:
This sounds Very good!  Thank you very much.

Is the data model for this lightweight to use with good API and manage progress/backlog/and so for "middle-tier" tasks?  Think of use-cases being served in old style database with asynchronous triggers following API calls, to change paradigm, like Kafka with consumer-groups.  Software agents can connect to couchdb's job-service api, claim ready to serve shard of queue for specific application-layer job,  get change-stream like replication queue and post notifications if needed to show progress.  It can be simpler and get good help from job infrastructure, make business infrastructure easier to control and manage.  They enjoy most of same things like indexing does, optional "reprocess whole database", live processing of newest records even during rebuild and so.

nickva commented:
@sansato

> Is the data model for this lightweight to use with good API and manage progress/backlog/and so for "middle-tier" tasks?

Ah, that's an interesting idea. The background jobs systems could in theory accept jobs from the outside if there was an HTTP API wrapper around it. So far in the design, jobs's keys and values are FDB tuples and/or json encoded objects so it seems doable. Perhaps something for a future iteration...

We also mention that workers could be implemented as any process which can connect to the FDB cluster. Generally it will be Erlang processes (or applications) but those workers could be written in other language as long as they can talk to the FDB cluster. And we might even end up with that when implementing some of the secondary indexing work. Having a completely custom set of workers, and a more fleshed out API for the workers to claim and do work would be possible. But I suggest we leave that for the future as well.

sansato commented:
for now, could highlight the integration spots to decrease odds that "future" resolves to "never".

nickva commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,287 @@
+periodically ping their health key. The period MUST be less than the
+`HealthTimeout` value. Other worker nodes, for example, one of the "neighbors"
+(to left or to the right) of the worker, will monitor that health entry. If a
+worker dies unexpectedly, the neighbor MUST remove all of the dead worker's
```
I mentioned that they should check their right neighbor.

rnewson commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+environment with a FoundationDB client. The only coordination between workers
+happens via the database. Workers can start and stop at any time. Workers
+monitor each other for liveliness and in case some workers abruptly terminate,
+all the jobs of a dead worker are re-enqueued into the global pending queue.
```
how do we detect if the job terminated successfully (job done) versus failed and therefore needs retrying?

nickva replied:
It's job type specific. Whether to re-enqueue the job to retry on error (with a backoff penalty) would be dependent on the type. This framework would only know if the job is running somewhere, is in a pending queue waiting to run, or has completed and will not be running anymore. 

The reason for the job being in pending state could be a retry or could be because it is a new job. The reason for completion could be a successful completion or failure.

I am ok with adding those extra states if it would generally work for indexing. It would work for replication (for reference https://docs.couchdb.org/en/stable/replication/replicator.html#replication-states)

So it could be:

  * "pending" : waiting to run in a queue
  * "running" : running (in a worker)
  * "completed" : successfully completed
  * "failed": permanent failure, job should not be re-inserted in the pending queue, user should delete it and re-create it
  * "error" (this also maps to "crashing" in the replicator case): An error has occurred but the job is pending for a retry. 

Should job retries and backoff schedule also be part of the jobs framework or each worker type should define its own policy. (ex. replication jobs use a doubling on each failed job start, with a max about 8 hours, what do indexing jobs do?)

garrensmith replied:
I think those states would work for indexing as well. I'm not sure if the retry and back off would have any value though.

nickva replied:
I update the RFC to refocus on jobs instead of workers instead.  But I didn't add the extra states to the job framework. It might make more sense for each job type to define its own. The job framework only cares about "pending", "running" and "finished". Those states are not even explicitly tracked as a state label but are derived based on other fields (if it is in the pending queue it is pending, if it was accepted by a worker it is "running").

I can see after implementing replication, indexing and others (couch-peruser) and seeing that all of them end up defining the same type-specific state, we could then move those to this framework

rnewson commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+
+`Job`: An unit of work, identified by a `JobId` and also having a `JobType`.
+
+`Job table`: A subspace holding the list of jobs indexed by `JobId`.
```
I think a `,` after 'jobs' would make this read better. `A subspace holding the list of jobs, indexed by JobId.`

rnewson commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+
+### Data model fields
+
+- `JobType`: The job type. It MAY be ``replication`` or ``indexing``, however,
```
inconsistent tense.

rnewson commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+- `CancelReq` : A boolean flag indicating a request to the worker to cancel the
+  job.
+
+- `JobOps` : Type-specific job options. Represented as a JSON object. This MAY
```
how is the JSON marshalled to fdb?

nickva replied:
As a binary. In Erlang it would be jiffy:encode(...)

rnewson commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+periodically ping their health key. The period MUST be less than the
+`HealthTimeout` value.
+
+Worker nodes MUST monitor their neighbors for liveness. When they detect an
```
this seems odd to me. Given the central registration, shouldn't the job of monitoring be done generically by some agent(s)? The act of updating a key to indicate liveness and another property indicating how out of date it can get seemed to indicate that monitoring was part of the framework and not the responsibility of everyone that uses the framework.

Also 'neighbour' was not defined before its first mention.

nickva replied:
Good point. I had flip-flopped in the previous version whether this should be job type specific (replication jobs know how to clean stale replication workers' jobs, indexers indexing jobs etc). But I think it can be done by the framework. So will update the wording there.

Also neighbour should be defined as a term perhaps. Will do that.

nickva replied:
I update the RFC to have a framework API where it is hopefully more clear where the monitoring happens.

rnewson commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+monitor their own presence in the membership list, and, if they detect they
+were removed, MUST stop running all their jobs. Then they MUST re-execute their
+startup logic: add themselves to the membership list, restart health key pings,
+and start picking jobs off of the pending jobs queue, and so on.
```
again, I'm lost. This implies that enforcing Mutual Exclusion is really the job of the workers to implement independently, in which case why have a framework?

nickva replied:
Good point. Update the RFC to contain a separate section regarding mutual exclusion.

sansato commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+worker has their own separate job queue. The advantage of that model is that
+the queues and jobs are sharded evenly amongst the workers, avoiding potential
+hot key ranges around the global queue pushes and pops. The disadvantage is that it
+seemed like a premature optimization. Also, it would only allow for a completely
```
This certain strategy prescribes premature, but sharded jobs could describe (additional job tuple term) parallelism opportunity, no reason workers must separately execute.  One worker can claim multiple jobs or ??

I don't see other sharding mentioned here, is other document concerning this with FDB?

nickva replied:
The sharding is happening in the current design (pre FDB) where each node is reading from it's own set of shards and running jobs from an in-memory queue (ets table).

> One worker can claim multiple jobs or ??

Right. The idea of worker is more of an application-type job manager. This manager would start multiple jobs of the same type, and then the would need to only maintain the liveness (health) of that supervisor or job manager. The equivalent would be something like the current replication scheduler.

However, after thinking about, it might be better to simplify and switch to a scheme with one worker being equivalent to a single process. So one worker executes on job only at a time...

janl commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+
+`Job`: An unit of work, identified by a `JobId` and also having a `JobType`.
+
+`Job table`: A subspace holding the list of jobs indexed by `JobId`.
```
```suggestion
`Job table`: A subspace holding the list of jobs, indexed by `JobId`.
```

janl commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+- `CancelReq` : A boolean flag indicating a request to the worker to cancel the
+  job.
+
+- `JobOps` : Type-specific job options. Represented as a JSON object. This MAY
```
```suggestion
- `JobOpts` : Type-specific job options. Represented as a JSON object. This MAY
```

janl commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+  contain details about a job's current state. It MAY have fields such as
+  `"update_seq"`, `"error_count"`, etc.
+
+- `HealthTimeout` : Worker timeout defined how often the worker will update
```
small bikeshed  on not using “health/alive/dead” terminology, maybe `ActivityTimeout`

nickva replied:
Hmm good point. Since we already have active_tasks as API. Jobs that update often enough are active. The ones that are not, end up as pending

janl commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+  they health entry to maintain liveness. The worker MUST update their health
+  entry at least this often to maintain liveness.
+
+- `HealthCounter` : Workers MUST updated this at least as often as the
```
small bikeshed  on not using “health/alive/dead” terminology, maybe `ActivityCounter`

nickva replied:
Agreed. Removed health from the RFC. `Health monitor` is now `Activity monitor`

janl commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+always sort before the normal ones. The uniqueness of `JobId` should reduce
+contention during insertion into the queue.
+
+Workers monitor the pending queue of their matching `JobType`. That is,
```
reference or example of how workers monitor their pending queue would be nice ✌️

nickva replied:
There is now an API description that hopefully make it more clear. See the an accept() function in the worker API and its implementation.

janl commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+changes using a watch.
+
+
+### Worker Liveness Monitoring
```
“Activeness” as per above comments.

janl commented on rfcs/007-background-workers.md:
```diff
@@ -0,0 +1,299 @@
+
+## Applications and Modules affected
+
+Replication and indexing
```
+ couch-peruser, and any other background/cleanup jobs we may or may not be introducing with the new data models.

nickva replied:
Good call.

nickva commented:
Thank you all for the feedback. Taking it into consideration, and also
discussing with @davisp I think it would make sense make these changes:

 1. Switch the focus from workers to jobs. Previously the focus was on workers.
   How many workers are running, how they are registered, monitoring their
   liveness, etc. The concept of worker was similar to a job manager or job
   supervisor. That introduced unecessary complexity. It was also confusing
   that on worker was running more than one job. So instead the API and the
   implementation switch to making a job as central concept. Users add job,
   resubmit job, etc. Worker now become just what function (or process) that
   happens to execute the job.

 2. Previously there was no clear distinction between how the framework would be
  implemented vs what framework API would look like. In the current version,
  the main part of the RFC focuses on the framework API and then implementation
  is discussed separately.

 3. Some concepts like "health" were removed and instead re-used "activity"
  since it is already something CouchDB knows about (thanks @janl for pointing
  it out).

 4. Based on indexing requirements, introduced the concept of job-resubmission.

 5. Mutual exclusion guarantees will be important for indexing and so there is
   a section discussing that aspect as well.

garrensmith commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,335 @@
+## Data Model
+
+`("couch_jobs", "data", Type, JobId) = (Sequence, WorkerLockId, JobOpts)`
+`("couch_jobs", "pending", Type, Priority, JobId) = ""`
```
Doesn't the pending queue job item need a Jobops section in the value space? How else would we define what the job is about?

nickva replied:
When the JobId is added to the pending queue it won't be removed from the main jobs table.

garrensmith commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,335 @@
+`accept()` :
+ * Generate a unique `WorkerLockId` UUID.
+ * Attempt to dequeue the item from the pending queue, then assign it the
+   `WorkerLockId` in the jobs table.
```
I'm a little confused on where the jobs table is is it this one:
`("couch_jobs", "data", Type, JobId) = (Sequence, WorkerLockId, JobOpts)`?

nickva replied:
Yeah that's the jobs table. Good point, I should have explicitly indicated that right below the data model section.

garrensmith commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,335 @@
+## Activity Monitor Implementation
+
+There is an activity monitor running for each job type. It performs periodic
+scans of the `"active"` range with sequences less than `"watches"` sequence for
```
Do you mean `activity` range? Otherwise I can't see in the data model where active range is.

nickva replied:
Good find, yeah should be the "activity" subspace. I'll fix it.

garrensmith commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,335 @@
+`update()`:
+ * Check if `WorkerLockId` matches, otherwise return `worker_conflict`
+ * Check if `"canceled"` field in the `JobOpts` is `true`, if so, return `canceled`
+ * Delete old `"activity"` sequence entry
```
This is interesting, what are the reasons for using a sequence to track this versus just an integer that we could atomically increment for during each update?

nickva replied:
Good question. I should develop the algorithm for the activity monitor a bit more.

The idea is that if the jobs are updated on-time, the activity check will return almost no results and will not need to keep anything in memory except the "watches" sequence for the from the previous check (`ActivityTimeout` seconds ago). With entries being bumped with atomic counters, we'd have to keep the whole job table in memory to compare the counters. I think both approaches might work for the purposes of activity monitoring, and if we find that there are significant performance issues, we could switch to using atomic counters. 

A few more details:

Every time a job is updating its "activity", it bumps the "watches" sequence. That makes "watches" sequence point to the newest updated sequence for that job type. That allows anyone, in this case the activity monitor, to remember the "watches" sequence value, then return some time later and ask "has anything not been updated since time I saw that sequence" by querying any entries *less* than its remembered sequence. If everything is working properly and jobs are updating, it should get an empty list.

This scheme also allows efficient watching of new change since the last remembered sequence , by querying anything *greater* than the last seen "watches" sequence. So it allows relatively efficient monitoring of jobs without having a separate watch for each job entry.

garrensmith replied:
Ok cool. Could you add that info to the RFC that explains it really nicely.

garrensmith commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,335 @@
+that type. The scan is performed as often as `ActivityTimeout`. The first time
+it scans the entries, it remembers the set of `JobId` and sequences it found.
+Subsequently, if those haven't changed since the last `ActivityTimeout` period,
+it will re-enqueue those jobs into the pending queue.
```
I think you might have to elaborate a bit more on the Activity monitor. What happens if the activity monitor crashes and forgets the sequence number it last checked? Is it one activity monitor per node or per job type or just one.

nickva replied:
Good call. I'll update the activity monitor section to both expand on how the algorithm work and list the steps separately as a bulleted list.

garrensmith commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,350 @@
+  * On subsequent runs:
+    * Call `get_range()` on `"activity"` entries less than the remembered sequence
+    * Read and store the new `"watches"` sequence
+    * If there are any stale entries returned, re-enqueue them
```
When a job is considered staled and needs to be re-enqueues we might to need call a job specific function. For indexes, only 1 worker can update an index at one time. So we will most likely have to create a lock in fdb so that no other workers try and update an index that is already being worked on. If the worker then dies, we would need to remove that lock otherwise that index will never be built. When the activity monitor re-enqueues it, it should also either send an event, or call some registered callback for when an indexing job fails with the job information so that their can be some job indexing clean up.

nickva replied:
Hopefully we can use the `WorkerLockId` and `worker_conflict` semantics. 

I wonder if this will work:

 * If indexing job update their activity on-time, then in the index update transaction, each job checks if `{"resubmit":true}` was set in the `JobOpts`.  If it is, it would call `index_cleanup(...)` and then call `finish(...)`. 

 * However it's possible the index updater job had failed to update its activity often enough so when calling `update(Tx, ...)` and it suddenly gets back `worker_conflict`. Then another worker may already be running the indexing job. Here I guess it depends if indexing cleanup for this stale job, can be performed while another (new) indexing job is already running. As far as the jobs framework is concerned, as long as indexing cleanup doesn't write to or update the "couch_jobs" that would work. But it would depend on the semantics of cleanup and if concurrent cleanup is allowed.

garrensmith replied:
Awesome idea, I think we could use `WorkerLockId` to indicate that a specific worker is building the index. Could we then make the JobId the same for for every build or update of a specific index? Then we can use that to know if an index is already being built and whether we need to add another pending job to update an index or just rely on the current job building the index.

nickva replied:
I think we could probably turn the JobId into a deterministic Id built from ```(Db, DbVersion, ViewGroup, View)```. That might help with matching the indexing job with the view. Then when we try to `add(...)` the job it will come back with a `{error, duplicate_job}` and then we could `resubmit()`.

nickva commented:
Prototype implementation. Sill WIP 

https://github.com/apache/couchdb/commit/f8c2411739082c4708f4b87b97d314a070d6f060

nickva commented:
Next WIP update

 https://github.com/apache/couchdb/commit/db4d1dc6ba2b7830fd8542033710652f1b787e35

This one has all the major pieces, including:
  - All jobs creation API: add(), remove(), submit(), get_job()
  - Workers API: finish(), resubmit(), update() and accept()
  - Directory path caching
  - Metadata changes check
  - Activity monitor
  - Type monitor (start notifiers and activity monitors when new types detected)
  - Job state subscription: subscribe(), unsubscribe()

nickva commented:
Latest update:

https://github.com/apache/couchdb/commit/f3036031adfb8744fa1218352032fc59ef627bbe

The basics now seem to work.

This is an example of subscribing to a job status and waiting for to finished.

```erlang
> couch_jobs:add(<<"t1">>, <<"j1">>, #{<<"data">> => #{my_data => blah}}).
ok

> f(Lock1), {ok, _, Lock1} = couch_jobs:accept(<<"t1">>).
{ok,<<"j1">>,<<"2be00fa90ec3ef5caaedb4d22d9d7f59">>}

> fabric2_fdb:transactional(fun(Tx) -> couch_jobs:update(Tx, <<"t1">>, <<"j1">>, #{data => #{my_data => <<"xyzc">>}}, Lock1) end).
ok

> fabric2_fdb:transactional(fun(Tx) -> couch_jobs:finish(Tx, <<"t1">>, <<"j1">>, #{}, Lock1) end).
ok
```

In a separate terminal

```erlang

> {ok, SubRef, InitState} = couch_jobs:subscribe(<<"t1">>, <<"j1">>).
{ok,{<0.1271.0>,#Ref<0.2568083611.2369781761.6545>},pending}

> couch_jobs:wait_job_state(SubRef, finished, infinity).
{<<"t1">>,<<"j1">>,finished}
```

nickva commented on rfcs/007-background-jobs.md:
````diff
@@ -0,0 +1,350 @@
+### Job Creation API ###
+
+```
+add(Type, JobId, JobOpts) -> ok | {error, Error}
````
There would be a duplicate job error returned.

nickva commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,350 @@
+developers using this framework to implement background jobs for indexing,
+replication, and couch-peruser.
+
+Both the job creation and the worker implementation APIs use a `JobOpts` map to
```
I can see it being the `Job` rather than `JobOpts`. I'll modify it.

The reason it was `JobOpts` is that technically a `JobOpts` could be an empty map / json object. `WorkerLockId` ends up as a separate tuple field. It's not something users can be expected to modify directly in the `Job` object, but they can, for instance, update a job's priority.

nickva commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,350 @@
+    the job. `Priority` is job-type specific.
+  * `"data"`: An opaque object (map), from the framework's point of view,
+    containing job-type specific data. It MAY contain an update sequence, or an
+    error message, for example.
```
Good call. I add the note that it must encode to JSON properly.

nickva commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,350 @@
+  * `"cancel"` : Boolean field defaulting to `false`. If `true` indicates the
+    user intends to stop a job's execution.
+  * `"resubmit"` : Boolean field defaulting to `false`. If `true` indicates
+    the job should be re-submitted.
```
Good point. Let's remove those fields from that section as users would modify those via the API. 

As for resubmit behavior, the API effectively sets a flag that the `finish(...)` implementation looks at. If it is set it will re-enqueue the job. So it's the `finish(...)` call that does the re-enqueuing. And it is either the job creation or the worker implementation (the processing) that can set the flat at any time before finish is called.

nickva commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,350 @@
+  * `"running"` : This job is currently running.
+  * `"finished"` : This job has finished running and is not pending.
+
+### Worker Implementation API
```
Yap, agree, `Job Processing API` sounds more clear.

nickva commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,350 @@
+
+### Worker Implementation API
+
+This API is to be used when implementing workers for various job types. The general pattern
```
A `Worker` is defined in the terminology section as:

 > `Worker` : A language-specific execution unit that runs the job. Could be an
Erlang process, a thread, or just a function

A job processor would be longer to type, but it's easy enough to replace, if it is seems more clear.

nickva commented on rfcs/007-background-jobs.md:
````diff
@@ -0,0 +1,350 @@
+ - Indicates that the job should be re-submitted for execution.
+
+```
+get_job(Type, JobId) -> {ok, JobOpts, JobState}
````
> Having this be `{ok, Job}` and then a `couch_job:state(Job)` function (along with accessors for other fields) seems like it'd be more user friendly especially given some comments I'm about to make down below.

Sounds reasonable, yeah.

> Also, `get_job` vs `accept` down below are slightly confusing. I assume `get_job` is intended more for monitoring and we should emphasize/enforce that a job description returned from this function must not be passed to the update functions below?

Yeah get_job is for monitoring. Will mentioned that the return value should not be used for updates. Or even better, just let users set the "data" field only via "accept".

nickva commented on rfcs/007-background-jobs.md:
````diff
@@ -0,0 +1,350 @@
+
+```
+accept(Type[, MaxPriority]) -> {ok, JobId, WorkerLockId} | not_found
+```
````
> Its not entirely clear to me how many simultaneous calls to `accept` should be made. There's a mention of a `job manager` but is that required or just a hint at a specific implementation? Is it possible to just spawn `N` consumers that all call `accept` independently?
> 
> Also does `accept` block? When is `not_found` returned? What should happen if `not_found` is returned? Do we sleep for a bit and try again? Is that an unrecoverable error like "I have no idea what `Type` you're talking about." sort of response?

Could spawn any number of consumers each calling `accept`. If any one can dequeue a job they'll return with a lock id, if they can't, they return `not_found`.  They maybe try again after some time. The job manager could be a gen_server that periodically calls accept, and maybe the more jobs it has already dequeued, the longer it sleeps in between accepts.

As the for type not being found, it's possible but it's also not a terminal condition in that it could be that there aren't any jobs of that type added yet. 

> Also, same for `get_job` it seems like it would be better to return `{ok, Job}` here to make client code more ergonomic so we're only carrying around one variable instead of the possible 5 that `finish` takes as arguments.

The `Job` object doesn't have the `JobId` in it. We could return the `Job` here but we definitely need the `JobId`. And since somewhere else we discussed maybe just letting users update the `data` field only, we might have less incentive to add the `Job` to the response.

nickva replied:
The latest version consolidates those fields into the `Job` opaque object, with functions to set `JobData` and resubmit flag.

nickva commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,350 @@
+represent a job. It MAY also contain these top level fields:
+
+  * `"priority"` : The value of this field will contain the `Priority` value of
+    the job. `Priority` is job-type specific.
```
The priority is simply a sorting subspace to putting some jobs ahead of others. It allows accepting some jobs first and not necessarily in the order in which they were added.

There are two extreme cases to consider:

1) There is no priority at all (which is the same as saying all jobs are the same `null` or `normal` priority). In this case the `queue` becomes a jobs `bucket`.  Any worker can pick any job regardless when it was added to the jobs pending queue. This setup would also minimize contention as a each could pick randomly pick one `JobId`

2) A strictly prioritized queue, with each Job having its own priority according the say a version stamp of when it was added.  Jobs are guaranteed to be accepted in the order in which they were added. But there will be more contention involved when dequeuing them.

The abstract priority as proposed allows having those extremes, and anything in between with a trade-off between ordering guarantees and dequeue contention.

Some examples:

Priority could be just two values "0-urgent" and "5-normal". There could be 10000 urgent and 10000 normal jobs.
```
 (..., "0-urgent", "36b40ba869104c3d84a5f6821c04d26d") = ...
 (..., "0-urgent", "c63dfe3a2f8f459da4ce9e444c15fc48") = ...
  ...
 (..., "5-normal", "61b5e60438cc4865a838c675ab1aa076") = ...
 (..., "5-normal", "aa8c364ba05149e29db3ab358d1203c8") = ...
```

Workers would first randomly pick through 10000 urgent ones and then only when those are accepted, start picking normal jobs.

Another case cold be with using timestamps with say a 5 second resolution vs 5 minute resolution. In one case there would be more contention, but the timing guarattee, but contention would lessen, because now, each worker could randomly choose to run any job from now till now + 5 minutes. There could now be thousands more jobs to pick from.

> seems like a bit of an overload on what I would traditionally consider as a "priority"

We could perhaps rename to something like "sort order" ? The replicator does need something prioritize some jobs over the others and it might be useful for indexing where say if there are 1000 design documents updated, we might choose to start indexing some of them first before others.

nickva commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,350 @@
+
+## Activity Monitor Implementation
+
+There is an activity monitor running for each job type. It performs periodic
```
Activity monitor is one of the definitions in the terminology section:

>  `Activity monitor` : Functionality implemented by the framework which checks
job liveness (activity). If job processor don't update their status often enough,
activity monitor will re-enqueue their jobs as pending. This ensures jobs make
progress even if some job processor terminate unexpectedly.

nickva commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,350 @@
+have both `completed` and `failed` as terminal states. In both cases the
+framework will consider those as `finished`, but job-type specific data MAY
+indicate that `completed` is finished with a successful result, and `failed` is
+finished with a terminal failure (say an invalid replication doc format).
```
> This discussion on job states is confusing to me. Referencing the data model I don't see a state variable anywhere except for the `pending` subspace. Which has me wondering where the running/complete/failed/finished etc states come into effect. I could see running as implicit by its inclusion in the `activity` subspaces. Though that also seems like it might be better off as `active` as the state name.

The states are implicit depending if they are pending or running. Running jobs have a worker lock id. I'll add the following to the `Job State Diagram` 

> These states are not tracked via an explicit state label. If the job is in the
pending queue, it is considered pending, if it has a lock id acquired by a job
processor, it is considered running.

How does that sound?

> I'm also not clear on what happens to a job after its "done" (abstractly, i.e., all of completed/failed/finished states). You've mentioned replications here as a reference which is a one approach since we maintain info about them so that users will realize why a replication isn't finished. But a secondary index update at some point needs to disappear from the database at some point right? Who makes that decision when? Or perhaps the "terminal" state here is something that's been pulled too far in from the replicator and maybe shouldn't be part of this RFC?

I'll remove the discussion about the terminal states. A job that is finished is not auto-removed, it stays in the jobs table. The job creator can inspect the result and remove it.  The worker maybe also remove the job when finished if the job wasn't re-enqueued.

garrensmith commented on rfcs/007-background-jobs.md:
````diff
@@ -0,0 +1,350 @@
+accept(Type[, MaxPriority]) -> {ok, JobId, WorkerLockId} | not_found
+```
+
+ - Dequeue a job from the pending queue. `WorkerLockId` is a UUID indicating
````
If we renaming workers to job processing, we should probably rename this to JobLockId to be consistent.

nickva replied:
Agree. Good idea

nickva commented on rfcs/007-background-jobs.md:
````diff
@@ -0,0 +1,350 @@
+   not execute any sooner than one hour from now.
+
+```
+finish(Tx, Type, JobId, JobOpts, WorkerLockId) -> ok | worker_conflict | canceled
````
Yeah, good idea, let's go with one `Job` object. 

I also like `halt` better than `worker_conflict`

nickva commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,350 @@
+   accepted. The intended usage is to allow `Priority` to be used as a way to
+   schedule job for executing at a future date. For example, `Priority` MAY
+   indicate that a replication job which has been repeatedly failing should
+   not execute any sooner than one hour from now.
```
The concept of `Priority` was removed and is not `ScheduledTime` parameter. This makes it more clear how it will be used, is less abstract and easier to reason about.

nickva commented:
@davisp, @garrensmith Thanks for the great comments and input. I had pushed the update that hopefully addresses most of the issues. I think the API looks cleaner and the structure of the RFC is more reasonable as well.

garrensmith commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,337 @@
+
+In the discussion below a job is considered to be an abstract unit of work. It
+is identified by a `JobId` and has a `JobType`. The user creates a job, and
+then it is executed by a job processor. A job processor is language-specific
```
`The user creates a job`, that is maybe not try. A user will perform some kind of action on CouchDB, create a replication doc or query an index. A CouchDB process would then create a job.

nickva replied:
The `user` is `client code`. It might be better to just use that instead of user.

garrensmith commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,337 @@
+
+### Job States
+
+Jobs in the system can be in a few different states. After a job is added and
```
I think you should say that jobs can be in 3 states.

garrensmith commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,337 @@
+Jobs in the system can be in a few different states. After a job is added and
+is waiting to run, the job is considered to be `pending`. A job executed by
+a job processor is considered to be `running`. When a job is neither `running`,
+nor `pending`, it is considered to be `finished`. This is the state transition
```
Should we mention that a finished job could be resubmitted later?

nickva replied:
If the job was running, then resubmitting it would make it pending again after the job processor calls `finish`. However if it already finished, then it won't start running again.

nickva replied:
It was updated such that job creator can just call `add/4,5` again. If the job is running it will be flagged to be re-enqueued when it finishes. If it is pending, only the scheduled time value may be updated (if provided). If the job was already finished it is re-enqueued for execution.

garrensmith commented on rfcs/007-background-jobs.md:
````diff
@@ -0,0 +1,337 @@
+set_type_timeout(Type, TimeoutSec) -> ok
+```
+
+ - Set the activity timeout for a job type.
````
Maybe we need to say that this needs to be set for a type before any jobs of that type can be processed.

nickva replied:
Yes, good call!

garrensmith commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,337 @@
+
+All the coordination between job creation and job processing is done via
+FoundationDB. There is a top level `"couch_jobs"` subspace. All the subspaces
+mentioned blow will be under this subspace.
```
s/blow/below

garrensmith commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,337 @@
+
+Each job managed by the framework will have an entry in the main `jobs table`.
+Pending jobs are added to a `pending queue` subspace. After they are
+accepted by a jobs processor, they jobs removed from the pending queue and added
```
The jobs are removed

garrensmith commented on rfcs/007-background-jobs.md:
```diff
@@ -0,0 +1,337 @@
+
+# Advantages and Disadvantages
+
+A previous draft of this RFC discussed an implementation centered around
```
I don't think this paragraph is really relevant. The advantages and disadvantages should be focused around this design.  I think the advantages are more along the lines that we have a central way to coordinate batch processing across a cluster. A nice unified api. Not sure around disadvantages

nickva replied:
Agree.

kocolosk commented:
@davisp @garrensmith @nickva is this one approved and ready to merge? I don't see any outstanding comments

nickva commented:
I think I addressed most of the issues. Didn't know if @davisp wanted to take another look at it

davisp commented:
I haven't got anything to add.
